### ISSUE TITLE
Bug: don't report build failure if network has problem

### ISSUE THREAD

**Issue by vatioz** [AnyStatus/Support#61]
We use anyStatus Desktop (v 1.4.101) for Jenkins builds, but I think this may be applied more generally.

Our use case:
We use VPN for connecting to our build server. Unfortunately VPN client is not able to maintain the connection in the case of any network disturbance and therefore we lose connection to build server very often (as we are changing network interfaces when (un)plugging to docking stations).

The problem:
In such case AnyStatus will report failure for all build jobs defined.
Not only this is very annoying as notifications will start blinking one after the other for good amount of time, but it is plain wrong. Network unavailability really **does not equal build failure**.

Solution:
My suggestion would be to use another state for network unavailability. Probably some gray icon or whatever. Also to not notify separately for each build when network is lost or revived. Rather don't show any notification in either case.

Note that I want to be notified about real build failures and stuff, so disabling notifications would not be an option here.

**Comment by AlonAm** [comment]
You're absolutely right, it is annoying.
I've been searching for a proper solution to detect such cases.
There's a difference between network unavailability and internet unavailability. 

Can you please describe what are the conditions in your case or how can it be detected?

**Comment by vatioz** [comment]
Hey,
what I would assume is that unless you get proper response from the server (jenkins) that would tell you whether the build succeeded or failed, you cannot tell what is the current state, hence _unknown_ state.

So if the request comes back, you can set **success** or **failure**.
If the request fails (i.e, 404, 500 HTTP responses or it times out, ...) you can **only** set **unknown** status.

Am I correct? What do you think?

**Comment by AlonAm** [comment]
That's an interesting approach.

So you suggest,

1. When an error occurs, set the status to Unknown instead of Error.
2. Bypass notifications when the status changes to Unkown.

Note, when a build fails, the status is Failed, not Error.

I think it can work...

**Comment by vatioz** [comment]
Oh, so there is already a difference? So there is FAIL state and there is ERROR state right now. Is that correct?
So maybe only some option "Notify error states" would suffice? 
Because what would be the difference between UNKNOWN and ERROR then? Are there usages for both?

Also, not only bypass notifications in the OK->UNKNOWN direction but in the opposite as well - UNKNOWN->OK. 
Although, this opposite direction may get a bit tricky, because there may be some failed builds when connection is revived. Do we want to be notified about them? Ultimate solution would be to have last states saved and then compare them with revived ones - and notify only about the changed ones.

So let's say I have 3 builds. Connection is good and 2 are passing, 1 is failing.
Connection goes down.
**No notification** about builds. (All builds are in either ERROR or UNKNOWN state.)
Connection goes up. Builds don't change, 2 are passing, 1 is failing, (the same ones).
**No notification** about builds.
Connection goes down.
**No notification**
Connection goes up after some time. Builds change - all failing now.
I would get **2 notifications** - I am interested only in the delta from the last known status ;)

**Comment by AlonAm** [comment]
There are no specific cases for Error except
we save last status, whatever that maybe
and skip notifications when the status change from Error to something else.
That works for some of the cases.

On Dec 19, 2017 3:18 PM, "Petr Kubát" <notifications@github.com> wrote:

Oh, so there is already a difference? So there is FAIL state and there is
ERROR state right now. Is that correct?
So maybe only some option "Notify error states" would suffice?
Because what would be the difference between UNKNOWN and ERROR then? Are
there usages for both?

Also, not only bypass notifications in the OK->UNKNOWN direction but in the
opposite as well - UNKNOWN->OK.
Although, this opposite direction may get a bit tricky, because there may
be some failed builds when connection is revived. Do we want to be notified
about them? Ultimate solution would be to have last states saved and then
compare them with revived ones - and notify only about the changed ones.

So let's say I have 3 builds. Connection is good and 2 are passing, 1 is
failing.
Connection goes down.
*No notification* about builds. (All builds are in either ERROR or UNKNOWN
state.)
Connection goes up. Builds don't change, 2 are passing, 1 is failing, (the
same ones).
*No notification* about builds.
Connection goes down.
*No notification*
Connection goes up after some time. Builds change - all failing now.
I would get *2 notifications* - I am interested only in the delta from the
last known status ;)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/AnyStatus/Support/issues/61#issuecomment-352746120>,
or mute
the thread

**Comment by AlonAm** [comment]
here's where notifications are created:
for builds: https://github.com/AnyStatus/API/blob/master/src/AnyStatus.API/Widgets/Build.cs
for everything else: https://github.com/AnyStatus/API/blob/master/src/AnyStatus.API/Notifications/DefaultNotificationFactory.cs

**Comment by AlonAm** [comment]
there's also a boolean property [here](https://github.com/AnyStatus/API/blob/5bc27fa25bc4586578a2ea3dac0b896344a277f7/src/AnyStatus.API/Widgets/Item.cs#L207) that indicates whether a notification is needed

```csharp
return ShowNotifications &&
           PreviousState != null &&
           PreviousState != State &&
           PreviousState != State.None;
```

**Comment by vatioz** [comment]
Well, it seems to me that what I am talking about matches current ERROR state. Skipping notification in transition from ERROR to different state is good. So the only missing thing would be

**option to skip notifications in transition from something to ERROR state**

That is enough to keep me happy (I think :) ).

**Comment by AlonAm** [comment]
great!
do you think it should be disabled by default?

**Comment by vatioz** [comment]
If this applies generally and not specifically to builds, I would leave the option disabled by default (keeping current behavior; showing notifications on error).

**Comment by AlonAm** [comment]
cool, ill add the new property (show error notifications) to all plugins, with True as default to keep current behavior, and override it for builds and releases (false) to apply the new behavior

**Comment by AlonAm** [comment]
for the record, there are other solutions such as monitoring network availability or monitoring a queue of notifications but for now lets try your solution and see how it goes :-)

**Comment by AlonAm** [comment]
Further down the road, the user will be able to configure how to  respond to status changes in different ways, like showing a notification or invoking some action

**Comment by AlonAm** [comment]
the new fix has been released in anystatus desktop 1.5.7
let's give it a try and see how it goes...

**Comment by vatioz** [comment]
Nice! That was fast! Thanks. I can confirm that I no longer get notifications for transitions to ERROR state. 

But, I get them when I get back on vpn (transition from ERROR to some other state). That I wasn't expecting.
![image](https://user-images.githubusercontent.com/14203428/34242352-5c8d9f18-e61b-11e7-8c72-c7fda4ab11de.png)

**Comment by AlonAm** [comment]
Hey there @SamuelEnglard,

In [TeamCity Plugin](https://github.com/AnyStatus/Plugins/blob/master/src/AnyStatus.Plugins/TeamCity/TeamCityBuild.cs#L83), you've modified the code to bypass notifications when the network connection is restored (thank you for that!).

Did it reduce the number annoying notifications when the network connection is restored?
Why did you choose to handle it only when State.Failed (build failed)?
Should'nt we completely bypass notifications when state changes from Error to something else?
thank you! :-)

```csharp
public override Notification CreateNotification()
        {
            if (State == State.Failed)
            {
                //todo: move "if" to API
                if (PreviousState == State.Error) return Notification.Empty; //bypass when network connection is restored.
                return new Notification($"{Name} failed. {StateText}", NotificationIcon.Error);
            }

            return base.CreateNotification();
}
```

**Comment by SamuelEnglard** [comment]
@AlonAm While I'd love to take credit for that, you did that one 😉 

AnyStatus/Plugins@3fa2f870a9660b77c91e817a3e867eba004cabef

My change was about showing status text on failure, so that's all I checked for.

AnyStatus/Plugins@4e9294bd0a09cb1c86007daa95c9c76ae7282a2a

**Comment by AlonAm** [comment]
@SamuelEnglard apologies, it must have been 2am ;-)

**Comment by SamuelEnglard** [comment]
@AlonAm it's ok, I write some of my best code at 2am!

**Comment by SamuelEnglard** [comment]
Though as I read through this thread I feel like this should be something handled by `Build` so that all get it "for free". There's no reason each sub class should be implementing this

**Comment by AlonAm** [comment]
You're right, i have added an option to show/hide error notifications for all widgets and set the default for builds to false

**Comment by AlonAm** [comment]
so I have released a new version (desktop 1.5.12) with the new behavior, including skipping notifications when errors recover.
lets give it a try and see how it goes...
thanks guys for the feedback <3

**Comment by vatioz** [comment]
I installed new version and I can confirm it is working as expected. Great!
